### PR TITLE
feat(iot-service): Add TokenCredential and AzureSasCredential support to Query and Twin clients

### PR DESCRIPTION
**Pull request by timtay-microsoft**
The twin client and the query clients are very intertwined, so I had to group them together for this PR

### CODE REVIEW

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperations.java [hunk 1]
```diff
@@ -214,9 +197,95 @@ public static HttpResponse request(
         request.setHeaderField(ACCEPT, ACCEPT_VALUE);
         request.setHeaderField(CONTENT_TYPE, ACCEPT_VALUE + "; " + ACCEPT_CHARSET);
 
+        if (payload != null)
+        {
+            request.setHeaderField(CONTENT_LENGTH, String.valueOf(payload.length));
```
Comment (bikamani) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperations.java:202]: 👍 
Is this optional and that's why did not exist before?

Comment (timtay-microsoft) [reply service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperations.java:202]: It is optional, but I'm surprised we don't send it anyways. It is very typical to include this header in every HTTP request

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 2]
```diff
@@ -126,31 +124,38 @@ public Query(int pageSize, QueryType requestQueryType) throws IllegalArgumentExc
...
-    }
 
-    /**
-     * Continuation token and page size to be used for next query request.
-     * @param continuationToken token to be used for next query request. Can be {@code null}.
-     * @param pageSize size batch for this query.
-     * @throws IOException if sending the request is unsuccessful because of input parameters.
-     * @throws IotHubException if sending the request is unsuccessful at the Hub.
-     */
-    private void continueQuery(String continuationToken, int pageSize) throws IOException, IotHubException
```
Comment (barustum) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:142]: this method is not being used?

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperations.java [hunk 3]
```diff
@@ -59,88 +62,68 @@ public static HttpResponse request(
...
             throw new IllegalArgumentException("HTTP Request timeout shouldn't not exceed " + timeoutInMs + DEFAULT_HTTP_TIMEOUT_MS + " milliseconds");
         }
 
-        /* Codes_SRS_DEVICE_OPERATIONS_21_006: [The request shall create a new SASToken with the ServiceConnect rights.] */
         String sasTokenString = new IotHubServiceSasToken(iotHubConnectionString).toString();
-        /* Codes_SRS_DEVICE_OPERATIONS_21_007: [If the SASToken is null or empty, the request shall throw IOException.] */
-         if((sasTokenString == null) || sasTokenString.isEmpty())
+         if (Tools.isNullOrEmpty(sasTokenString))
         {
             throw new IOException("Illegal sasToken null or empty");
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperations.java:94]: Not IOException, change to whatever is appropriate in java.

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperations.java [hunk 4]
```diff
@@ -170,23 +153,23 @@ public static HttpResponse request(
...
+        if (method == null)
         {
             throw new IllegalArgumentException("Http requests must provide a non-null http method");
         }
 
         String sasTokenString = new IotHubServiceSasToken(iotHubConnectionString).toString();
-        if((sasTokenString == null) || sasTokenString.isEmpty())
+        if (Tools.isNullOrEmpty(sasTokenString))
         {
             throw new IOException("Illegal sasToken null or empty");
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperations.java:174]: Not IOException. Change to IllegalArgumentException

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/RawTwinQuery.java [hunk 5]
```diff
@@ -137,9 +218,24 @@ public synchronized String next(Query query) throws IOException, IotHubException
...
+    }
+
+    private String getAuthenticationToken()
+    {
+        // Three different constructor types for this class, and each type provides either a TokenCredential implementation,
+        // an AzureSasCredential instance, or just the connection string. The sas token can be retrieved from the non-null
+        // one of the three options.
+        if (this.credential != null)
+        {
+            return this.credential.getToken(new TokenRequestContext()).block().getToken();
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/RawTwinQuery.java:232]: Check with service team add add the necessary prefix and suffix - https://github.com/Azure/azure-sdk-for-net/blob/8970711b5c82c9fac3f7135529ab57de5a294f08/sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs#L218

File: service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperationsTest.java [hunk 6]
```diff
@@ -651,7 +619,7 @@ public void invokeSucceed(
                 httpRequest.setReadTimeoutMillis(DEFAULT_HTTP_TIMEOUT_MS);
                 times = 1;
                 httpRequest.setHeaderField(anyString, anyString);
-                times = 5;
+                times = 4;
```
Comment (vinagesh) [inline service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperationsTest.java:622]: Why this change?

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceTwin.java [hunk 7]
```diff
@@ -538,4 +659,21 @@ private DeviceTwinDevice jsonToDeviceTwinDevice(String json) throws IOException
...
     }
+
+    private String getAuthenticationToken()
+    {
+        // Three different constructor types for this class, and each type provides either a TokenCredential implementation,
+        // an AzureSasCredential instance, or just the connection string. The sas token can be retrieved from the non-null
+        // one of the three options.
+        if (this.credential != null)
+        {
+            return this.credential.getToken(new TokenRequestContext()).block().getToken();
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceTwin.java:670]: Confirm with service team on what prefix and suffix to add.
https://github.com/Azure/azure-sdk-for-net/blob/8970711b5c82c9fac3f7135529ab57de5a294f08/sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs#L218

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 8]
```diff
@@ -167,18 +172,18 @@ private void continueQuery(String continuationToken, int pageSize) throws IOExce
...
-    public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionString,
-                                   URL url,
-                                   HttpMethod method,
-                                   Long timeoutInMs) throws IOException, IotHubException
+    public QueryResponse sendQueryRequest(
+            IotHubConnectionString iotHubConnectionString,
+            URL url,
+            HttpMethod method,
+            Long timeoutInMs)
+            throws IOException, IotHubException
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:180]: Do we really get a IOException?

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 9]
```diff
@@ -260,12 +258,14 @@ public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionStr
...
-                                          int httpReadTimeout,
-                                          Proxy proxy) throws IOException, IotHubException
+    public QueryResponse sendQueryRequest(
+            IotHubConnectionString iotHubConnectionString,
+            URL url,
+            HttpMethod method,
+            int httpConnectTimeout,
+            int httpReadTimeout,
+            Proxy proxy)
+            throws IOException, IotHubException
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:268]: Why IOException in non deprecated method?

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 10]
```diff
@@ -335,13 +335,192 @@ public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionStr
...
+    {
+        this.url = url;
+        this.httpMethod = method;
+        this.azureSasCredential = azureSasCredential;
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:362]: Same

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 11]
```diff
@@ -335,13 +335,192 @@ public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionStr
...
+        {
+            QueryRequestParser requestParser = new QueryRequestParser(this.query);
+            payload = requestParser.toJson().getBytes();
+        }
+        else
+        {
+            payload = new byte[0];
+        }
+
+        HttpResponse httpResponse = DeviceOperations.request(azureSasCredential.getSignature(), url, method, payload, null, httpConnectTimeout, httpReadTimeout, proxy);
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:390]: nit: Split args to multiple lines

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 12]
```diff
@@ -335,13 +335,192 @@ public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionStr
...
+                    this.responseQueryType = QueryType.fromString(header.getValue());
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        if (this.responseQueryType == null || this.responseQueryType == QueryType.UNKNOWN)
+        {
+            throw new IOException("Query response type is not defined by IotHub");
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:411]: Why IOException? Change to IoTHubException

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 13]
```diff
@@ -335,13 +335,192 @@ public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionStr
...
+        }
+
+        if (this.requestQueryType != this.responseQueryType)
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:414]: same

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 14]
```diff
@@ -335,13 +335,192 @@ public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionStr
...
+     * @throws IotHubException If HTTP response other then status ok is received.
+     */
+    public QueryResponse sendQueryRequest(
+            TokenCredential credential,
+            URL url,
+            HttpMethod method,
+            int httpConnectTimeout,
+            int httpReadTimeout,
+            Proxy proxy)
+            throws IOException, IotHubException
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:443]: Why IOException?

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java [hunk 15]
```diff
@@ -335,13 +335,192 @@ public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionStr
...
+                    this.responseQueryType = QueryType.fromString(header.getValue());
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        if (this.responseQueryType == null || this.responseQueryType == QueryType.UNKNOWN)
+        {
+            throw new IOException("Query response type is not defined by IotHub");
```
Comment (vinagesh) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:505]: Change this and like 510 to IoThubExeption

File: service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceTwinTest.java [hunk 16]
```diff
@@ -208,7 +200,7 @@ public void getTwinOperationSucceeds(@Mocked DeviceTwinDevice mockedDevice, @Moc
                 mockedHttpRequest.setReadTimeoutMillis(anyInt);
                 times = 1;
                 mockedHttpRequest.setHeaderField(anyString, anyString);
-                times = 5;
+                times = 4;
```
Comment (vinagesh) [inline service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceTwinTest.java:203]: why this change?

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java (hunk 2 above)
Comment (timtay-microsoft) [reply service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:142]: It wasn't being used, yeah

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java (hunk 9 above)
Comment (timtay-microsoft) [reply service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:268]: This method sends an HTTP request, which can throw an IOException

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java (hunk 14 above)
Comment (timtay-microsoft) [reply service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:443]: This method sends an HTTP request which may throw an IOException

File: service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceTwinTest.java (hunk 16 above)
Comment (timtay-microsoft) [reply service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceTwinTest.java:203]: Some of the mocking in the unit test broke when I switched to validating the requestId parameter by calling Tools.isNullOrEmpty, so the header case for adding a request Id header isn't called from this unit test anymore

File: service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperationsTest.java (hunk 6 above)
Comment (timtay-microsoft) [reply service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperationsTest.java:622]: Some of the mocking in the unit test broke when I switched to validating the requestId parameter by calling Tools.isNullOrEmpty, so the header case for adding a request Id header isn't called from this unit test anymore

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java (hunk 9 above)
Comment (vinagesh) [reply service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java:277]: I see. I missed that the method in line 170 was already deprecated and this was not to remove the IOException.

**Review by vinagesh — APPROVED** [review]
